### DISCUSSION TITLE
Why does API Route Handler not have access to `res` and API Route Middleware not return a `(req, res)`?

### DISCUSSION THREAD

**Discussion by darwin403** [vercel/next.js#50344] [App Router]
### Summary

The earlier handlers (pre-appDir) allowed one to manage the `req` and `res` states from the handler and hence build custom `withCacheResponse`, `withSensitiveDetails` kinda middlewares that handle business logic by modifing both the request and the response before finally sending it over to the handler. However, with the new `appDir`, I can no longer access the response variable from the previous step in my route handler, nor can I pass the response object from the middleware onto the handler. **This is a signifcant problem, since one can no longer pass a state from the middleware to the handler.**

In principle, the abstraction should be such that for a API "Middleware"  and "Handler" should transform the tuples as follows. The current architecture and what the ideal are given below

```js
// Current Architecture

// middelware.ts
const middleware = (req) => newRes`

// route.ts
const handler = (req) => finalRes
```

```js
// Ideal Architecture

// middelware.ts
const middleware = (req,res) => (newReq, newRes)`

// route.ts
const handler = (req,res) => finalRes
```

But I think in the interest of a single `middleware.ts` file for both pages and routes, this fuinctionality has now been trivialised with a `response` from middleware.

I know I can build some custom abstraction for my project. But this seems to be the defacto standard with the API routes all over the industry like express.js/ I think the `middleware.ts` for the frontend pages and backend routes should function differently and preferably should be managable in separate unit files for the API routes atleast. Perhaps, the move to merge all types of middleware.ts into one file might work for content oriented websites. But large-scale applications would require a better API middleware handling.

Would urge the authors to please re-consider this if it makes sense. Thanks!

**Comment by sametcodes** [comment]
Agreed. Probably the purpose was splitting the middleware and handler function in order to make them possible to deploy as serverless functions. But it's not ideal. It would be great to specify if it's serverless or serverful and it could work according to that.

**Comment by ethan-krich** [comment]
You could import NextResponse. `import NextResponse from "next/server"`

**Reply by sametcodes** [reply]
That's not the point. The point is the middleware and handler functions do not share the same context and eventually it doesn't allow you to pass values like `res.locals.user` to handlers. `NextResponse` is can be used only to send responses.
